R/best_case_worst_case_transition_risk_profile_at_company_level.R

```diff
@@ -0,0 +1,88 @@
+best_case_worst_case_transition_risk_profile_at_company_level <- function(data) {
+  product <- data |>
+    unnest_product()
+
+  avg_best_case_worst_case_at_product_level <- product |>
+    create_avg_best_case_worst_case_at_product_level() |>
+    prepare_for_join_at_company_level()
+
+  company <- data |>
+    unnest_company() |>
+    left_join(avg_best_case_worst_case_at_product_level, by = c(col_companies_id(),
+      "benchmark_tr_score_avg" = col_transition_risk_grouped_by()
+    )) |>
+    polish_best_case_worst_case_transition_risk_at_company_level()
+
+  tilt_profile(nest_levels(product, company))
+}
+
+create_avg_best_case_worst_case_at_product_level <- function(data) {
+  crucial_cols <- c(
+    col_companies_id(), col_europages_product(),
+    col_transition_risk_grouped_by(), col_transition_risk_category(),
+    "amount_of_distinct_products"
+  )
+  check_crucial_cols(data, crucial_cols)
+
+  data |>
+    select(crucial_cols) |>
+    count_min_risk_category_per_company_benchmark() |>
+    count_max_risk_category_per_company_benchmark() |>
+    add_avg_transition_risk_best_case() |>
+    add_avg_transition_risk_worst_case()
+}
+
+count_min_risk_category_per_company_benchmark <- function(data) {
+  mutate(data,
+    n_min_risk_category_per_company_benchmark =
+      sum_transition_risk_categories(.data, c("low", "medium", "high")),
+    .by = c(col_companies_id(), col_transition_risk_grouped_by())
+  )
+}
+
+count_max_risk_category_per_company_benchmark <- function(data) {
+  mutate(data,
+    n_max_risk_category_per_company_benchmark =
+      sum_transition_risk_categories(.data, c("high", "medium", "low")),
+    .by = c(col_companies_id(), col_transition_risk_grouped_by())
+  )
+}
+
+add_avg_transition_risk_best_case <- function(data) {
+  mutate(data,
+    avg_transition_risk_best_case =
+      (.data$n_min_risk_category_per_company_benchmark /
+        .data$amount_of_distinct_products)
+  )
+}
+
+add_avg_transition_risk_worst_case <- function(data) {
+  mutate(data,
+    avg_transition_risk_worst_case =
+      (.data$n_max_risk_category_per_company_benchmark /
+        .data$amount_of_distinct_products)
+  )
+}
+
+sum_transition_risk_categories <- function(data, risk_order) {
+  sum(data$transition_risk_category ==
+    risk_first_occurance(
+      data,
+      col_transition_risk_category(),
+      risk_order
+    ), na.rm = TRUE)
```

Comment:
It seems a bit odd that the function operates both with columns (`data$col`) and also with the entire dataframe (`data`). This may be a sign of sub-optimal design.

Reply:
@maurolepore I am not sure how I can improve this. Please have a look at the function where I used it:
```
count_max_risk_category_per_company_benchmark <- function(data) {
  mutate(data,
    n_max_risk_category_per_company_benchmark =
      sum_transition_risk_categories(.data, c("high", "medium", "low")),
    .by = c(col_companies_id(), col_transition_risk_grouped_by())
  )
}
```

Here I passed `.data` as an input to `sum_transition_risk_categories`, and please note that `risk_first_occurance` also expects the `.data` argument as input for its correct functioning. Am I not doing it right still?

Reply:
I checkout out the code but I'm struggling to fully understand what's going on. In general, I find it a lil surprising that this code is creating a column from a call to a function that operates on the entire dataset. But if your tests yield what you expect, then refactoring is of relatively low importance.